# 0. Import Library

In [1]:
import pandas as pd
import requests 
from bs4 import BeautifulSoup

# 1. Data Collecting

In [2]:
url = 'https://www.ft.com/ft1000-2022'
r = requests.get(url)
soup = BeautifulSoup(r.text, 'html.parser')
js = pd.read_html(str(soup.find_all('table')))

In [3]:
df = pd.DataFrame(js[0])
df

Rank          Name in 2021 ranking in 2020 ranking    Country  \
0        1.0       Swappie              No              No    Finland   
1        2.0   Kilo Health              No              No  Lithuania   
2        3.0           OCI              No              No         UK   
3        4.0      OnlyFans              No              No         UK   
4        5.0         Enpal              No              No    Germany   
...      ...           ...             ...             ...        ...   
996    997.0     Digitalpa              No              No      Italy   
997    998.0  Faktenkontor              No              No    Germany   
998    999.0        CLAREO             Yes             Yes     France   
999   1000.0    Laca Trade             Yes             Yes      Italy   
1000     NaN           NaN             NaN             NaN        NaN   

                     Sector  Absolute Growth Rate %  \
0                Technology                19152.82   
1                    Health                16541.73   
2        Financial Services                13133.13   
3                Technology                11928.05   
4                    Energy                11441.27   
...                     ...                     ...   
996   Management Consulting                  154.82   
997             Advertising                  154.81   
998                  Retail                  154.80   
999                Property                  154.60   
1000                    NaN                     NaN   

      Compound Annual Growth Rate (CAGR) %  Revenue 2020 (€)  \
0                                   477.43        97611814.0   
1                                   450.05        57318766.0   
2                                   409.59       568322073.0   
3                                   393.63       316732986.0   
4                                   386.88        56109613.0   
...                                    ...               ...   
996                                  36.59         1731340.0   
997                                  36.59        30967000.0   
998                                  36.58        18854708.0   
999                                  36.55        24741510.0   
1000                                   NaN               NaN   

      Revenue 2017 (€)  ...  Unnamed: 990 Unnamed: 991  Unnamed: 992  \
0             507000.0  ...           NaN          NaN           NaN   
1             344428.0  ...           NaN          NaN           NaN   
2            4325512.0  ...           NaN          NaN           NaN   
3            2652185.0  ...           NaN          NaN           NaN   
4             486165.0  ...           NaN          NaN           NaN   
...                ...  ...           ...          ...           ...   
996           679440.0  ...           NaN          NaN           NaN   
997         12153143.0  ...           NaN          NaN           NaN   
998          7399729.0  ...           NaN          NaN           NaN   
999          9717860.0  ...           NaN          NaN           NaN   
1000               NaN  ...           NaN          NaN           NaN   

      Unnamed: 993  Unnamed: 994  Unnamed: 995  Unnamed: 996  Unnamed: 997  \
0              NaN           NaN           NaN           NaN           NaN   
1              NaN           NaN           NaN           NaN           NaN   
2              NaN           NaN           NaN           NaN           NaN   
3              NaN           NaN           NaN           NaN           NaN   
4              NaN           NaN           NaN           NaN           NaN   
...            ...           ...           ...           ...           ...   
996            NaN           NaN           NaN           NaN           NaN   
997            NaN           NaN           NaN           NaN           NaN   
998            NaN           NaN           NaN           NaN           NaN   
999            NaN           NaN           NaN          

In [4]:
df.to_csv('data/FT1000_v0.csv', index = False)

### Về Dataset:  Danh sách 1000 công ty tăng trưởng nhanh nhất Châu Âu 

Dataset được scrape từ bài báo `Special Report FT 1000: Europe's Fastest Growing Companies` ở trang `Financial Times`, một nhật báo ở lĩnh vực kinh tế tài chính. Báo cáo này cho thấy sự thay đổi trong sự tăng trưởng của các công ty trong giai đoạn Covid 2020/

Những công ty được chọn là những công ty đã phát triển đủ để vượt qua giai doạn khó khăn bởi Covid, Brexit và đóng băng chuỗi cung ứng toàn cầu. 

#### Phương pháp luận thu thập dữ liệu

`FT 1000: Europe's Fastest Growing Companies` là danh sách gồm 1000 công ty ở Châu Âu đạt mức trăng trưởng doanh thu cao nhất trong giai đoạn 2017 - 2020. Tuy nhiên danh sách này cũng không thể hiện sự tăng trưởng thực tế của tất cả các công ty ở Châu Âu nho có một vài công ty không tham gia khảo sát vì vấn đề bảo mật thông tin cũng như một vài lí do khác.

Dự án được tiến hành từ 1/10/2021 tới 31/12/2021 thông qua việc điền form khảo sát online và trên giấy, cho phép các công ty đăng kí thông qua website của `Statista` và `Financial Times`. Ngoài ra thông qua nghiên cứu các dữ liệu công khai, `Statista` xác định top 10000 công ty tiềm năng trở thành top 1000 và mời các công ty này tham gia.

Các thông số về doanh thu phải được xác nhận bởi các giám đốc điều hành, giám đốc tài chính và các thành viên khác trong ban điều hành của công ty. Đối với những công ty chỉ với 3 thành viên hoặc ít hơn, doanh thu của họ sẽ được kiểm tra thêm

#### Điều kiện để trở thành top 1000

Để lọt vào danh sách các công ty phát triển nhanh nhất Châu Âu, một công ty cần thỏa các điều kiện sau:
- Doanh thu ít nhất 100,000 Euro (hoặc đơn vị tiền tệ tương đương) năm 2017 
- Doanh thu ít nhất 1.5 triệu Euro (hoặc đơn vị tiền tệ tương đương) năm 2020
- Là công ty độc lập (không phải công ty con hoặc chi nhánh văn phòng của công ty khác dưới bất kì hình thức nào)
- Nếu công ty được lên sàn chứng khoán, giá cổ phiếu không được giảm từ 50% trở lên kể từ năm 2019
- Một vài điều kiện về doanh thu khác mà nhóm chưa dịch được

#### Công thức tính tỉ lệ tăng trưởng

- Tỉ lệ tăng trưởng kép hàng năm (Compound Annual Growth Rate) - CAGR: 

$$
CAGR = \frac{\text{Doanh thu năm 2020}}{\text{Doanh thu năm 2017}} ^ \frac{1}{3}
$$

- Tỉ lệ tăng trưởng tuyệt đối (Absolute growth):

$$
\frac{\text{Doanh thu năm 2020}}{\text{Doanh thu năm 2017}} - 1
$$

Ý nghĩa của các tỉ lệ này sẽ được giải thích ở phần Data Preprocessing

#### Đánh giá và đảm báo chất lượng dữ liệu
Tất cả dữ liệu được báo cáo bởi các công ty đã được xử lí và kiểm tra bởi Statista. Các trường dữ liệu khuyết đều được nghiên cứu kĩ lưỡng. Các công ti không đạt đủ điều kiện nêu ở trên sẽ bị loại. Tỉ lệ tăng trưởng tuyệt đối thấp nhất để được xếp hạng ở đợt này (2021) là 36.5%